app/util/event_log.py

```diff
@@ -0,0 +1,265 @@
+        The time is considered stale if it is is older than start_time + LOG_CACHE_EXPIRE_TIME_IN_HOURS
+
+        :rtype: bool
+        """
+        if len(self._log_cache) == 0:
+            return False
+        else:
+            latest = self._latest_timestamp_in_cache()
+            oldest = self._oldest_timestamp_in_cache()
+            return oldest + LOG_CACHE_EXPIRE_TIME_IN_HOURS * 60 * 60 < latest
+
+    def _latest_timestamp_in_cache(self):
+        """
+        :rtype: float | None
+        """
+        if len(self._log_cache) == 0:
+            return None
+        else:
+            return self._log_cache[-1]['__timestamp__']
+
+    def _oldest_timestamp_in_cache(self):
+        """
+        :rtype: float | None
+        """
+        if len(self._log_cache) == 0:
+            return None
+        else:
+            return self._log_cache[0]['__timestamp__']
+
+    def get_events(self, since_timestamp=None, since_id=None):
+        """
+        Retrieve all events from the current eventlog since the given timestamp or event id.
+        This is used to expose events via the API and is useful for building dashboards
+        that monitor the system.
+
+        :param since_timestamp: Get all events after (greater than) this timestamp
+        :type since_timestamp: float | None
+        :param since_id:  Get all events after (greater than) this id.
+        :type since_id: int | None
+        :return: The list of events in the given range
+        :rtype: list[dict] | None
+        """
+        if self.logging_disabled or self.filename.upper() == 'STDOUT':
+            return None
+
+        if since_timestamp is not None and since_id is not None:
+            raise ValueError('since_timestamp and since_id can not be used at the same time')
+        events = []
+        condition_met = False
+        if self._should_try_get_event_from_log_cache():
+            # use the events from cache, but failover to fetch from the file
+            events, condition_met = self._get_events_from_reversed_generator(
+                generator=self._reversed_log_cache_event_generator(),
+                since_timestamp=since_timestamp,
+                since_id=since_id
+            )
+
+        if not condition_met or not events:
+            # if we didnt get back events from the cache, use the filesystem as
+            # as the source of truth
+            events, condition_met = self._get_events_from_reversed_generator(
+                generator=self._reversed_log_file_event_generator(),
+                since_timestamp=since_timestamp,
+                since_id=since_id
+            )
+
+        return events
+
+    def _should_try_get_event_from_log_cache(self, since_id=None,
+                                             since_timestamp=None):
+        if since_id is None and since_timestamp is None:
+            return True
+        elif since_timestamp and since_timestamp <= self._oldest_timestamp_in_cache():
```

Comment:
Just to make sure we're on the same page -- I think that if we keep the option for the user to specify either id or timestamp, then right here we need to check both the id **and** the timestamp to determine if the cache has all the data we want since we'll only ever have at most one of id or timestamp values specified.